# Pandas 

(If you're using the code files, please open pandas_lessons.py)

## The importance of data preprocessing

Data preprocessing (also called data wrangling, cleaning, scrubbing, etc) is the most important thing you will do with your data because it sets the stage for the analysis part of your data analysis workflow. The preprocessing you do largely depends on what kind of data you have, what sort of analysis you'll be doing with your data, and what you intend to do with the results.

Preprocessing is also a process for getting to know your data, and can answer questions such as these (and more): 

- What kind of data are you working with? 
- Is it categorical, continuous, or a mix of both? 
- What's the distribution of features in your dataset? 
- What sort of wrangling do you have to do?
- Do you have any missing data? 
- Do you need to remove missing data?
- Do you need only a subset of your data?
- Do you need more data?
- Or less?

The questions you'll have to answer are, again, dependent upon the data that you're working with, and preprocessing can be a way to figure that out.

## What is Pandas?

Pandas is by far my favorite preprocessing tool. It's a data wrangling/modeling/analysis tool that is similar to R and Excel; in fact, the DataFrame data structure in Pandas was named after the DataFrame in R. Pandas comes with several easy-to-use data structures, two of which (the `Series` and the `DataFrame`) I'll be covering here.

I'll also be covering a bunch of different wrangling tools, as well as a couple of analysis tools.

## Why Pandas?

So, why would you want to use Python, as opposed to tools like R and Excel? I like to use it because I like to keep everything in Python, from start to finish. It just makes it easier if I don't have to switch back and forth between other tools. Also, if I have to build in preprocessing as part of a production system, which I've had to do at my job, it makes sense to just do it in Python from the beginning. 

Pandas is great for preprocessing, as we'll see, and it can be easily combined with other modules from the scientific Python stack.

## Pandas data structures

Pandas has several different data structures, but we're going to talk about the `Series` and the `DataFrame`.

### The Series

The `Series` is a one-dimensional array that can hold a variety of data types, including a mix of those types. The row labels in a `Series` are collectively called the index. You can create a `Series` in a few different ways. Here's how you'd create a `Series` from a list.

In [1]:
import pandas as pd

some_numbers = [2, 5, 7, 3, 8]

series_1 = pd.Series(some_numbers)
series_1

0    2
1    5
2    7
3    3
4    8
dtype: int64

To specify an index, you can also pass in a list.

In [2]:
ind = ['a', 'b', 'c', 'd', 'e']

series_2 = pd.Series(some_numbers, index=ind)
series_2

a    2
b    5
c    7
d    3
e    8
dtype: int64

We can pull that index back out again, too, with the `.index` attribute.

In [3]:
series_2.index

Index(['a', 'b', 'c', 'd', 'e'], dtype='object')

You can also create a `Series` with a dictionary. The keys of the dictionary will be used as the index, and the values will be used as the `Series` array.

In [4]:
more_numbers = {'a': 9, 'b': 'eight', 'c': 7.5, 'd': 6}

series_3 = pd.Series(more_numbers)
series_3

a        9
b    eight
c      7.5
d        6
dtype: object

Notice how, in that previous example, I created a `Series` with integers, a float, and a string.

### The DataFrame

The `DataFrame` is Pandas' most used data structure. It's a two and greater dimensional structure that can also hold a variety of mixed data types. It's similar to a spreadsheet in Excel or a SQL table. You can create a `DataFrame` with a few different methods. First, let's look at how to create a `DataFrame` from multiple `Series` objects.

In [5]:
combine_series = pd.DataFrame([series_2, series_3])
combine_series

a      b    c  d    e
0  2      5  7.0  3  8.0
1  9  eight  7.5  6  NaN

Notice how in column `b`, we have two kinds of data. If a column in a `DataFrame` contains multiple types of data, the data type (or `dtype`) of the column will be chosen to accomodate all of the data. We can look at the data types of different columns with the `.dtypes` attribute. `object` is the most general, which is what has been chosen for column `b`.

In [6]:
combine_series.dtypes

a      int64
b     object
c    float64
d      int64
e    float64
dtype: object

Another way to create a `DataFrame` is with a dictionary of lists. This is pretty straightforward:

In [7]:
data = {'col1': ['i', 'love', 'pandas', 'so', 'much'],
        'col2': ['so', 'will', 'you', 'i', 'promise']}

df = pd.DataFrame(data)
df

col1     col2
0       i       so
1    love     will
2  pandas      you
3      so        i
4    much  promise

## File I/O

It's really easy to read data into Pandas from a file. Pandas will read your file directly into a `DataFrame`. There are multiple ways to read in files, but they all work in the same way. Here's how you read in a CSV file:

In [8]:
wine = pd.read_csv('../data/wine.csv')
wine.head()

abv  malic_acid   ash  alcalinity  magnesium  total_phenols  flavanoids  \
0  14.23        1.71  2.43        15.6        127           2.80        3.06   
1  13.20        1.78  2.14        11.2        100           2.65        2.76   
2  13.16        2.36  2.67        18.6        101           2.80        3.24   
3  14.37        1.95  2.50        16.8        113           3.85        3.49   
4  13.24        2.59  2.87        21.0        118           2.80        2.69   

   nonflavanoid_phenols  proanthocyacins  color   hue  dilution  proline  \
0                  0.28             2.29   5.64  1.04      3.92     1065   
1                  0.26             1.28   4.38  1.05      3.40     1050   
2                  0.30             2.81   5.68  1.03      3.17     1185   
3                  0.24             2.18   7.80  0.86      3.45     1480   
4                  0.39             1.82   4.32  1.04      2.93      735   

   wine_type  
0          1  
1          1  
2          1  
3          1  
4          1

Reading in a text file is just as easy. Make sure to pass in `'\t'` to the delimiter parameter.

In [9]:
auto_mpg = pd.read_csv('../data/auto_mpg.txt', delimiter='\t')
auto_mpg.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model  \
0  18.0          8         307.0        130    3504          12.0     70   
1  15.0          8         350.0        165    3693          11.5     70   
2  18.0          8         318.0        150    3436          11.0     70   
3  16.0          8         304.0        150    3433          12.0     70   
4  17.0          8         302.0        140    3449          10.5     70   

   origin                   car_name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

## Exploring the data

Here are some different ways to explore the data we have. Let's first take a look at some of the basic characteristics of the auto_mpg dataset. You can easily find the number of rows and the number of columns a dataframe has using the `.shape` attribute.

In [10]:
auto_mpg.shape

(398, 9)

You've already seen the `head()` function, which returns the first five lines in the dataset. To grab the last 5 lines, you can use the `tail()` function:

In [11]:
auto_mpg.tail()

mpg  cylinders  displacement horsepower  weight  acceleration  model  \
393  27.0          4         140.0         86    2790          15.6     82   
394  44.0          4          97.0         52    2130          24.6     82   
395  32.0          4         135.0         84    2295          11.6     82   
396  28.0          4         120.0         79    2625          18.6     82   
397  31.0          4         119.0         82    2720          19.4     82   

     origin         car_name  
393       1  ford mustang gl  
394       2        vw pickup  
395       1    dodge rampage  
396       1      ford ranger  
397       1       chevy s-10

Getting column names from a `DataFrame` is also easy and can be done using the `.columns` attribute.

In [13]:
wine.columns

Index(['abv', 'malic_acid', 'ash', 'alcalinity', 'magnesium', 'total_phenols',
       'flavanoids', 'nonflavanoid_phenols', 'proanthocyacins', 'color', 'hue',
       'dilution', 'proline', 'wine_type'],
      dtype='object')

Another useful thing you can do is generate some summary statistics using the `describe()` function. The `describe()` function calculates descriptive statistics like the mean, standard deviation, and quartile values for continuous and integer data that exist in your dataset. Don't worry, Pandas won't try to calculate the standard deviation of your categorical values!

In [14]:
wine.describe()

abv  malic_acid         ash  alcalinity   magnesium  \
count  178.000000  178.000000  178.000000  178.000000  178.000000   
mean    13.000618    2.336348    2.366517   19.494944   99.741573   
std      0.811827    1.117146    0.274344    3.339564   14.282484   
min     11.030000    0.740000    1.360000   10.600000   70.000000   
25%     12.362500    1.602500    2.210000   17.200000   88.000000   
50%     13.050000    1.865000    2.360000   19.500000   98.000000   
75%     13.677500    3.082500    2.557500   21.500000  107.000000   
max     14.830000    5.800000    3.230000   30.000000  162.000000   

       total_phenols  flavanoids  nonflavanoid_phenols  proanthocyacins  \
count     178.000000  178.000000            178.000000       178.000000   
mean        2.295112    2.029270              0.361854         1.590899   
std         0.625851    0.998859              0.124453         0.572359   
min         0.980000    0.340000              0.130000         0.410000   
25%         1.742500    1.205000              0.270000         1.250000   
50%         2.355000    2.135000              0.340000         1.555000   
75%         2.800000    2.875000              0.437500         1.950000   
max         3.880000    5.080000              0.660000         3.580000   

            color         hue    dilution      proline   wine_type  
count  178.000000  178.000000  178.000000   178.000000  178.000000  
mean     5.058090    0.957449    2.611685   746.893258    1.938202  
std      2.318286    0.228572    0.709990   314.907474    0.775035  
min      1.280000    0.480000    1.270000   278.000000    1.000000  
25%      3.220000    0.782500    1.937500   500.500000    1.000000  
50%      4.690000    0.965000    2.780000   673.500000    2.000000  
75%      6.200000    1.120000    3.170000   985.000000    3.000000  
max     13.000000    1.710000    4.000000  1680.000000    3.000000

Another useful thing you can do to explore your data is to sort it. Let's say we wanted to sort our `auto_mpg DataFrame` by mpg. This is very easy as well:

In [16]:
#auto_mpg.sort(columns='mpg').tail()

auto_mpg.sort_values(by=['mpg']).tail()

mpg  cylinders  displacement horsepower  weight  acceleration  model  \
326  43.4          4          90.0         48    2335          23.7     80   
394  44.0          4          97.0         52    2130          24.6     82   
325  44.3          4          90.0         48    2085          21.7     80   
329  44.6          4          91.0         67    1850          13.8     80   
322  46.6          4          86.0         65    2110          17.9     80   

     origin              car_name  
326       2    vw dasher (diesel)  
394       2             vw pickup  
325       2  vw rabbit c (diesel)  
329       3   honda civic 1500 gl  
322       3             mazda glc

## Lesson: let's see what's going on in our data!

This dataset is data on credit approvals. The column names and data were changed to protect the confidentiality of the data.

In [17]:
f = pd.read_csv('../data/credit_approval.csv')

f.columns

# How do you read in that file?

# Can you grab just the column names?


Index(['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N',
       'O', 'P'],
      dtype='object')

In [18]:
# How many rows and columns does the dataframe have?
f.shape

(690, 16)

In [19]:
# Now, look at the first 5 lines
f.head()

A      B      C  D  E  F  G     H  I  J  K  L  M    N    O  P
0  b  30.83  0.000  u  g  w  v  1.25  t  t  1  f  g  202    0  +
1  a  58.67  4.460  u  g  q  h  3.04  t  t  6  f  g   43  560  +
2  a   24.5  0.500  u  g  q  h  1.50  t  f  0  f  g  280  824  +
3  b  27.83  1.540  u  g  w  v  3.75  t  t  5  t  g  100    3  +
4  b  20.17  5.625  u  g  w  v  1.71  t  f  0  f  s  120    0  +

In [20]:
# Now, look at the last 5 lines
f.tail()

A      B       C  D  E   F   G     H  I  J  K  L  M    N    O  P
685  b  21.08  10.085  y  p   e   h  1.25  f  f  0  f  g  260    0  -
686  a  22.67   0.750  u  g   c   v  2.00  f  t  2  t  g  200  394  -
687  a  25.25  13.500  y  p  ff  ff  2.00  f  t  1  t  g  200    1  -
688  b  17.92   0.205  u  g  aa   v  0.04  f  f  0  f  g  280  750  -
689  b     35   3.375  u  g   c   h  8.29  f  f  0  t  g    0    0  -

In [21]:
# Can you describe() the data? (Notice how Pandas only "describes" the numerical data!)
f.describe()

C           H          K              O
count  690.000000  690.000000  690.00000     690.000000
mean     4.758725    2.223406    2.40000    1017.385507
std      4.978163    3.346513    4.86294    5210.102598
min      0.000000    0.000000    0.00000       0.000000
25%      1.000000    0.165000    0.00000       0.000000
50%      2.750000    1.000000    0.00000       5.000000
75%      7.207500    2.625000    3.00000     395.500000
max     28.000000   28.500000   67.00000  100000.000000

In [23]:
# Let's sort on column H
f.sort_values(by=['H'])

A      B       C  D  E   F   G     H  I  J   K  L  M    N      O  P
351  b  22.17   0.585  y  p  ff  ff   0.0  f  f   0  f  g  100      0  -
633  b  32.42   2.165  y  p   k  ff   0.0  f  f   0  f  g  120      0  -
456  b  34.58   0.000  ?  ?   ?   ?   0.0  f  f   0  f  p    ?      0  -
585  b  73.42  17.750  u  g  ff  ff   0.0  t  f   0  t  g    0      0  +
584  a  28.08  15.000  y  p   e   z   0.0  t  f   0  f  g    0  13212  +
..  ..    ...     ... .. ..  ..  ..   ... .. ..  .. .. ..  ...    ... ..
586  b  64.08  20.000  u  g   x   h  17.5  t  t   9  t  g    0   1000  +
212  b  60.08  14.500  u  g  ff  ff  18.0  t  t  15  t  g    0   1000  +
250  b  40.25  21.500  u  g   e   z  20.0  t  t  11  f  g    0   1200  +
221  b  65.42  11.000  u  g   e   z  20.0  t  t   7  t  g   22      0  +
44   b  56.42  28.000  y  p   c   v  28.5  t  t  40  f  g    0     15  +

[690 rows x 16 columns]

## Working with dataframes

Pandas has a ton of functionality for manipulating and wrangling the data. Let's look at a bunch of different ways to select and subset our data.

### Selecting columns and rows

There are multiple ways to select by both rows and columns. From index to slicing to label to position, there are a variety of methods to suit your data wrangling needs.

Let's select just the mpg column from the `auto_mpg DataFrame`. This works similar to how you would access values from a dictionary:

In [24]:
auto_mpg['mpg']

0      18.0
1      15.0
2      18.0
3      16.0
4      17.0
       ... 
393    27.0
394    44.0
395    32.0
396    28.0
397    31.0
Name: mpg, Length: 398, dtype: float64

You can do exactly the same thing by using mpg as an attribute:

In [25]:
auto_mpg.mpg

0      18.0
1      15.0
2      18.0
3      16.0
4      17.0
       ... 
393    27.0
394    44.0
395    32.0
396    28.0
397    31.0
Name: mpg, Length: 398, dtype: float64

To extract rows from a `DataFrame`, you can use the slice method, similar to how you would slice a list. Here's how we would grab rows 7-13 from the wine `DataFrame`:

In [26]:
wine[7:14]

abv  malic_acid   ash  alcalinity  magnesium  total_phenols  flavanoids  \
7   14.06        2.15  2.61        17.6        121           2.60        2.51   
8   14.83        1.64  2.17        14.0         97           2.80        2.98   
9   13.86        1.35  2.27        16.0         98           2.98        3.15   
10  14.10        2.16  2.30        18.0        105           2.95        3.32   
11  14.12        1.48  2.32        16.8         95           2.20        2.43   
12  13.75        1.73  2.41        16.0         89           2.60        2.76   
13  14.75        1.73  2.39        11.4         91           3.10        3.69   

    nonflavanoid_phenols  proanthocyacins  color   hue  dilution  proline  \
7                   0.31             1.25   5.05  1.06      3.58     1295   
8                   0.29             1.98   5.20  1.08      2.85     1045   
9                   0.22             1.85   7.22  1.01      3.55     1045   
10                  0.22             2.38   5.75  1.25      3.17     1510   
11                  0.26             1.57   5.00  1.17      2.82     1280   
12                  0.29             1.81   5.60  1.15      2.90     1320   
13                  0.43             2.81   5.40  1.25      2.73     1150   

    wine_type  
7           1  
8           1  
9           1  
10          1  
11          1  
12          1  
13          1

Pandas also has tools for purely label-based selection of rows and columns using the `.loc` indexer. The `.loc` indexer takes input as `[row, column]`. 

For example, let's say we wanted to select the abv value in the 8th instance in our wine `DataFrame`:

In [27]:
wine.loc[8,'abv']

14.83

We can also use `.loc` to grab slices. It's important to note that `.loc` interprets the index as a *label*. This means that, if we select a range, it will grab the last item in the range, unlike slicing in a list. The index is the label for the rows. Let's grab the abv for rows 8 to 11 from the wine `DataFrame`.

In [28]:
wine.loc[8:11, 'abv']

8     14.83
9     13.86
10    14.10
11    14.12
Name: abv, dtype: float64

And, as you might expect, we can select multiple columns by passing in a list of column names. Let's also grab ash and color for rows 8 to 11.

In [29]:
wine.loc[8:11, ['abv', 'ash', 'color']]

abv   ash  color
8   14.83  2.17   5.20
9   13.86  2.27   7.22
10  14.10  2.30   5.75
11  14.12  2.32   5.00

Finally, let's just grab all columns for rows 8 to 11.

In [30]:
wine.loc[8:11, :]

abv  malic_acid   ash  alcalinity  magnesium  total_phenols  flavanoids  \
8   14.83        1.64  2.17        14.0         97           2.80        2.98   
9   13.86        1.35  2.27        16.0         98           2.98        3.15   
10  14.10        2.16  2.30        18.0        105           2.95        3.32   
11  14.12        1.48  2.32        16.8         95           2.20        2.43   

    nonflavanoid_phenols  proanthocyacins  color   hue  dilution  proline  \
8                   0.29             1.98   5.20  1.08      2.85     1045   
9                   0.22             1.85   7.22  1.01      3.55     1045   
10                  0.22             2.38   5.75  1.25      3.17     1510   
11                  0.26             1.57   5.00  1.17      2.82     1280   

    wine_type  
8           1  
9           1  
10          1  
11          1

So, `.loc` provides functionality for a very specific and precise selection method.

Pandas has tools for purely position-based selection of rows and columns using the `.iloc` indexer, which works exactly how slicing a list works. The `.iloc` indexer also takes input as `[row, column]`, but takes only integer input. If we wanted to access the 60th row and the model value from `auto_mpg`, it would look like this (remember that integer indexing is 0-based):

In [31]:
auto_mpg.iloc[60, 6]

72

To grab rows 60-63 and the last three columns from the `auto_mpg DataFrame`, we would need to do the following:

In [32]:
auto_mpg.iloc[60:64, 6:9]

model  origin             car_name
60     72       1       chevrolet vega
61     72       1  ford pinto runabout
62     72       1     chevrolet impala
63     72       1     pontiac catalina

`.iloc` again works like slicing a list, based on position, so it does not grab the last item, like `.loc` does.

To grab all values and those last three columns from the `auto_mpg DataFrame`:

In [33]:
auto_mpg.iloc[:, 6:9]

model  origin                   car_name
0       70       1  chevrolet chevelle malibu
1       70       1          buick skylark 320
2       70       1         plymouth satellite
3       70       1              amc rebel sst
4       70       1                ford torino
..     ...     ...                        ...
393     82       1            ford mustang gl
394     82       2                  vw pickup
395     82       1              dodge rampage
396     82       1                ford ranger
397     82       1                 chevy s-10

[398 rows x 3 columns]

One of my favorite methods for selecting data is through boolean indexing. Boolean indexing is similar to the WHERE clause in SQL in that it allows you to filter out data based on certain criteria. Let's see how this works.

Let's select from the wine `DataFrame` where `wine_type` is type 1.

In [34]:
wine[wine['wine_type'] == 1]

abv  malic_acid   ash  alcalinity  magnesium  total_phenols  flavanoids  \
0   14.23        1.71  2.43        15.6        127           2.80        3.06   
1   13.20        1.78  2.14        11.2        100           2.65        2.76   
2   13.16        2.36  2.67        18.6        101           2.80        3.24   
3   14.37        1.95  2.50        16.8        113           3.85        3.49   
4   13.24        2.59  2.87        21.0        118           2.80        2.69   
5   14.20        1.76  2.45        15.2        112           3.27        3.39   
6   14.39        1.87  2.45        14.6         96           2.50        2.52   
7   14.06        2.15  2.61        17.6        121           2.60        2.51   
8   14.83        1.64  2.17        14.0         97           2.80        2.98   
9   13.86        1.35  2.27        16.0         98           2.98        3.15   
10  14.10        2.16  2.30        18.0        105           2.95        3.32   
11  14.12        1.48  2.32        16.8         95           2.20        2.43   
12  13.75        1.73  2.41        16.0         89           2.60        2.76   
13  14.75        1.73  2.39        11.4         91           3.10        3.69   
14  14.38        1.87  2.38        12.0        102           3.30        3.64   
15  13.63        1.81  2.70        17.2        112           2.85        2.91   
16  14.30        1.92  2.72        20.0        120           2.80        3.14   
17  13.83        1.57  2.62        20.0        115           2.95        3.40   
18  14.19        1.59  2.48        16.5        108           3.30        3.93   
19  13.64        3.10  2.56        15.2        116           2.70        3.03   
20  14.06        1.63  2.28        16.0        126           3.00        3.17   
21  12.93        3.80  2.65        18.6        102           2.41        2.41   
22  13.71        1.86  2.36        16.6        101           2.61        2.88   
23  12.85        1.60  2.52        17.8         95           2.48        2.37   
24  13.50        1.81  2.61        20.0         96           2.53        2.61   
25  13.05        2.05  3.22        25.0        124           2.63        2.68   
26  13.39        1.77  2.62        16.1         93           2.85        2.94   
27  13.30        1.72  2.14        17.0         94           2.40        2.19   
28  13.87        1.90  2.80        19.4        107           2.95        2.97   
29  14.02        1.68  2.21        16.0         96           2.65        2.33   
30  13.73        1.50  2.70        22.5        101           3.00        3.25   
31  13.58        1.66  2.36        19.1        106           2.86        3.19   
32  13.68        1.83  2.36        17.2        104           2.42        2.69   
33  13.76        1.53  2.70        19.5        132           2.95        2.74   
34  13.51        1.80  2.65        19.0        110           2.35        2.53   
35  13.48        1.81  2.41        20.5        100           2.70        2.98   
36  13.28        1.64  2.84        15.5        110           2.60        2.68   
37  13.05        1.65  2.55        18.0         98           2.45        2.43   
38  13.07        1.50  2.10        15.5         98           2.40        2.64   
39  14.22        3.99  2.51        13.2        128           3.00        3.04   
40  13.56        1.71  2.31        16.2        117           3.15        3.29   
41  13.41        3.84  2.12        18.8         90           2.45        2.68   
42  13.88        1.89  2.59        15.0        101           3.25        3.56   
43  13.24        3.98  2.29        17.5        103           2.64        2.63   
44  13.05        1.77  2.10        17.0        107           3.00        3.00   
45  14.21        4.04  2.44        18.9        111           2.85        2.65   
46  14.38        3.59  2.28        16.0        102           3.25        3.17   
47  13.90        1.68  2.12        16.0        101           3.10        3.39   
48  14.10        2.02  2.40        18

This works with any comparison operators, like >, < >=, !=, and so on. For example, we can select everything from the wine `DataFrame` where the value in the magnesium column is less than 100.

In [35]:
wine[wine['magnesium'] < 100]

abv  malic_acid   ash  alcalinity  magnesium  total_phenols  \
6    14.39        1.87  2.45        14.6         96           2.50   
8    14.83        1.64  2.17        14.0         97           2.80   
9    13.86        1.35  2.27        16.0         98           2.98   
11   14.12        1.48  2.32        16.8         95           2.20   
12   13.75        1.73  2.41        16.0         89           2.60   
..     ...         ...   ...         ...        ...            ...   
170  12.20        3.03  2.32        19.0         96           1.25   
171  12.77        2.39  2.28        19.5         86           1.39   
172  14.16        2.51  2.48        20.0         91           1.68   
173  13.71        5.65  2.45        20.5         95           1.68   
177  14.13        4.10  2.74        24.5         96           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyacins     color   hue  \
6          2.52                  0.30             1.98  5.250000  1.02   
8          2.98                  0.29             1.98  5.200000  1.08   
9          3.15                  0.22             1.85  7.220000  1.01   
11         2.43                  0.26             1.57  5.000000  1.17   
12         2.76                  0.29             1.81  5.600000  1.15   
..          ...                   ...              ...       ...   ...   
170        0.49                  0.40             0.73  5.500000  0.66   
171        0.51                  0.48             0.64  9.899999  0.57   
172        0.70                  0.44             1.24  9.700000  0.62   
173        0.61                  0.52             1.06  7.700000  0.64   
177        0.76                  0.56             1.35  9.200000  0.61   

     dilution  proline  wine_type  
6        3.58     1290          1  
8        2.85     1045          1  
9        3.55     1045          1  
11       2.82     1280          1  
12       2.90     1320          1  
..        ...      ...        ...  
170      1.83      510          3  
171      1.63      470          3  
172      1.71      660          3  
173      1.74      740          3  
177      1.60      560          3  

[97 rows x 14 columns]

You can also say 'not' with the tilde: ~

Let's select from the wine `DataFrame` where magnesium is NOT less than 100, which is equivalent to saying greater than or equal to.

In [36]:
wine[~(wine['magnesium'] < 100)]

abv  malic_acid   ash  alcalinity  magnesium  total_phenols  \
0    14.23        1.71  2.43        15.6        127           2.80   
1    13.20        1.78  2.14        11.2        100           2.65   
2    13.16        2.36  2.67        18.6        101           2.80   
3    14.37        1.95  2.50        16.8        113           3.85   
4    13.24        2.59  2.87        21.0        118           2.80   
..     ...         ...   ...         ...        ...            ...   
168  13.58        2.58  2.69        24.5        105           1.55   
169  13.40        4.60  2.86        25.0        112           1.98   
174  13.40        3.91  2.48        23.0        102           1.80   
175  13.27        4.28  2.26        20.0        120           1.59   
176  13.17        2.59  2.37        20.0        120           1.65   

     flavanoids  nonflavanoid_phenols  proanthocyacins  color   hue  dilution  \
0          3.06                  0.28             2.29   5.64  1.04      3.92   
1          2.76                  0.26             1.28   4.38  1.05      3.40   
2          3.24                  0.30             2.81   5.68  1.03      3.17   
3          3.49                  0.24             2.18   7.80  0.86      3.45   
4          2.69                  0.39             1.82   4.32  1.04      2.93   
..          ...                   ...              ...    ...   ...       ...   
168        0.84                  0.39             1.54   8.66  0.74      1.80   
169        0.96                  0.27             1.11   8.50  0.67      1.92   
174        0.75                  0.43             1.41   7.30  0.70      1.56   
175        0.69                  0.43             1.35  10.20  0.59      1.56   
176        0.68                  0.53             1.46   9.30  0.60      1.62   

     proline  wine_type  
0       1065          1  
1       1050          1  
2       1185          1  
3       1480          1  
4        735          1  
..       ...        ...  
168      750          3  
169      630          3  
174      750          3  
175      835          3  
176      840          3  

[81 rows x 14 columns]

It's also possible to combine these boolean indexers. Make sure you enclose them in parentheses. This is something I usually forget.

Let's select from wine where magnesium is less than 100 and the type of wine is type 1.

In [37]:
wine[(wine['magnesium'] < 100) & (wine['wine_type'] == 1)]

abv  malic_acid   ash  alcalinity  magnesium  total_phenols  flavanoids  \
6   14.39        1.87  2.45        14.6         96           2.50        2.52   
8   14.83        1.64  2.17        14.0         97           2.80        2.98   
9   13.86        1.35  2.27        16.0         98           2.98        3.15   
11  14.12        1.48  2.32        16.8         95           2.20        2.43   
12  13.75        1.73  2.41        16.0         89           2.60        2.76   
13  14.75        1.73  2.39        11.4         91           3.10        3.69   
23  12.85        1.60  2.52        17.8         95           2.48        2.37   
24  13.50        1.81  2.61        20.0         96           2.53        2.61   
26  13.39        1.77  2.62        16.1         93           2.85        2.94   
27  13.30        1.72  2.14        17.0         94           2.40        2.19   
29  14.02        1.68  2.21        16.0         96           2.65        2.33   
37  13.05        1.65  2.55        18.0         98           2.45        2.43   
38  13.07        1.50  2.10        15.5         98           2.40        2.64   
41  13.41        3.84  2.12        18.8         90           2.45        2.68   
50  13.05        1.73  2.04        12.4         92           2.72        3.27   
51  13.83        1.65  2.60        17.2         94           2.45        2.99   

    nonflavanoid_phenols  proanthocyacins  color   hue  dilution  proline  \
6                   0.30             1.98   5.25  1.02      3.58     1290   
8                   0.29             1.98   5.20  1.08      2.85     1045   
9                   0.22             1.85   7.22  1.01      3.55     1045   
11                  0.26             1.57   5.00  1.17      2.82     1280   
12                  0.29             1.81   5.60  1.15      2.90     1320   
13                  0.43             2.81   5.40  1.25      2.73     1150   
23                  0.26             1.46   3.93  1.09      3.63     1015   
24                  0.28             1.66   3.52  1.12      3.82      845   
26                  0.34             1.45   4.80  0.92      3.22     1195   
27                  0.27             1.35   3.95  1.02      2.77     1285   
29                  0.26             1.98   4.70  1.04      3.59     1035   
37                  0.29             1.44   4.25  1.12      2.51     1105   
38                  0.28             1.37   3.70  1.18      2.69     1020   
41                  0.27             1.48   4.28  0.91      3.00     1035   
50                  0.17             2.91   7.20  1.12      2.91     1150   
51                  0.22             2.29   5.60  1.24      3.37     1265   

    wine_type  
6           1  
8           1  
9           1  
11          1  
12          1  
13          1  
23          1  
24          1  
26          1  
27          1  
29          1  
37          1  
38          1  
41          1  
50          1  
51          1

If you wanted to, you could just keep on chaining the booleans together. Let's add on where the abv is greater than 14.

In [38]:
wine[(wine['magnesium'] < 100) & (wine['wine_type'] == 1) & (wine['abv'] > 14)]

abv  malic_acid   ash  alcalinity  magnesium  total_phenols  flavanoids  \
6   14.39        1.87  2.45        14.6         96           2.50        2.52   
8   14.83        1.64  2.17        14.0         97           2.80        2.98   
11  14.12        1.48  2.32        16.8         95           2.20        2.43   
13  14.75        1.73  2.39        11.4         91           3.10        3.69   
29  14.02        1.68  2.21        16.0         96           2.65        2.33   

    nonflavanoid_phenols  proanthocyacins  color   hue  dilution  proline  \
6                   0.30             1.98   5.25  1.02      3.58     1290   
8                   0.29             1.98   5.20  1.08      2.85     1045   
11                  0.26             1.57   5.00  1.17      2.82     1280   
13                  0.43             2.81   5.40  1.25      2.73     1150   
29                  0.26             1.98   4.70  1.04      3.59     1035   

    wine_type  
6           1  
8           1  
11          1  
13          1  
29          1

Another method of selecting data is using the `isin()` function. If you pass in a list to `isin()`, it will return a `DataFrame` of booleans. True means that the value at that index is in the list you passed into `isin()`.

Let's take the first five rows of the `auto_mpg DataFrame` and check for certain values existing in the `DataFrame`.

In [39]:
auto_mpg_5 = auto_mpg.head()

vals = [8, 150, 12.0, 'ford torino']
auto_mpg_5.isin(vals)

mpg  cylinders  displacement  horsepower  weight  acceleration  model  \
0  False       True         False       False   False          True  False   
1  False       True         False       False   False         False  False   
2  False       True         False       False   False         False  False   
3  False       True         False       False   False          True  False   
4  False       True         False       False   False         False  False   

   origin  car_name  
0   False     False  
1   False     False  
2   False     False  
3   False     False  
4   False      True

If it says `True`, it means that one of the values from the `vals` list occurs there.

## Lesson: let's try some of these on some data!

In [41]:
# Extract column C from the credit_approval dataframe we read in above
f.C

0       0.000
1       4.460
2       0.500
3       1.540
4       5.625
        ...  
685    10.085
686     0.750
687    13.500
688     0.205
689     3.375
Name: C, Length: 690, dtype: float64

In [44]:
# Slice rows 5-10 from the credit_approval dataframe
f[5:11]

A      B       C  D  E   F  G      H  I  J  K  L  M    N      O  P
5   b  32.08   4.000  u  g   m  v  2.500  t  f  0  t  g  360      0  +
6   b  33.17   1.040  u  g   r  h  6.500  t  f  0  t  g  164  31285  +
7   a  22.92  11.585  u  g  cc  v  0.040  t  f  0  f  g   80   1349  +
8   b  54.42   0.500  y  p   k  h  3.960  t  f  0  f  g  180    314  +
9   b   42.5   4.915  y  p   w  v  3.165  t  f  0  t  g   52   1442  +
10  b  22.08   0.830  u  g   c  h  2.165  f  f  0  t  g  128      0  +

In [50]:
# How would you look up the value for the 13th row in column C by label (loc)?
f.loc[13,'C']
f[13:14]

A      B     C  D  E  F  G     H  I  J  K  L  M  N     O  P
13  b  48.08  6.04  u  g  k  v  0.04  f  f  0  f  g  0  2690  +

In [51]:
# How would you look up the same thing by position (iloc)?
f.iloc[13,2]

6.04

In [58]:
# What if I wanted to select all data from credit_approval based on column C being greater than 5?
f[f['C'] > 5]

A      B       C  D  E   F   G      H  I  J   K  L  M    N     O  P
4    b  20.17   5.625  u  g   w   v  1.710  t  f   0  f  s  120     0  +
7    a  22.92  11.585  u  g  cc   v  0.040  t  f   0  f  g   80  1349  +
12   a  38.25   6.000  u  g   k   v  1.000  t  f   0  t  g    0     0  +
13   b  48.08   6.040  u  g   k   v  0.040  f  f   0  f  g    0  2690  +
14   a  45.83  10.500  u  g   q   v  5.000  t  t   7  t  g    0     0  +
..  ..    ...     ... .. ..  ..  ..    ... .. ..  .. .. ..  ...   ... ..
676  a  30.58  10.665  u  g   q   h  0.085  f  t  12  t  g  129     3  -
677  b  19.42   7.250  u  g   m   v  0.040  f  t   1  f  g  100     1  -
678  a  17.92  10.210  u  g  ff  ff  0.000  f  f   0  f  g    0    50  -
685  b  21.08  10.085  y  p   e   h  1.250  f  f   0  f  g  260     0  -
687  a  25.25  13.500  y  p  ff  ff  2.000  f  t   1  t  g  200     1  -

[232 rows x 16 columns]

In [62]:
# What if I wanted to select data based on column C being greater than 5 and column F being equal to 'w'?
f[(f['C']>5 ) & (f['F']=='W')]

Empty DataFrame
Columns: [A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P]
Index: []

In [63]:
# What if I wanted to look at a boolean DataFrame of where values are in ['t', 's', 100, 0] in credit_approval?
v=['t','s',100.0]

f.isin(v)

A      B      C      D      E      F      G      H      I      J  \
0    False  False  False  False  False  False  False  False   True   True   
1    False  False  False  False  False  False  False  False   True   True   
2    False  False  False  False  False  False  False  False   True  False   
3    False  False  False  False  False  False  False  False   True   True   
4    False  False  False  False  False  False  False  False   True  False   
..     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
685  False  False  False  False  False  False  False  False  False  False   
686  False  False  False  False  False  False  False  False  False   True   
687  False  False  False  False  False  False  False  False  False   True   
688  False  False  False  False  False  False  False  False  False  False   
689  False  False  False  False  False  False  False  False  False  False   

         K      L      M      N      O      P  
0    False  False  False  False  False  False  
1    False  False  False  False  False  False  
2    False  False  False  False  False  False  
3    False   True  False  False  False  False  
4    False  False   True  False  False  False  
..     ...    ...    ...    ...    ...    ...  
685  False  False  False  False  False  False  
686  False   True  False  False  False  False  
687  False   True  False  False  False  False  
688  False  False  False  False  False  False  
689  False   True  False  False  False  False  

[690 rows x 16 columns]

## Groupby

`groupby()` is just like SQL's 'group by' clause. What groupby does is a three-step process:

- Split the data
- Apply a function to the split groups
- Recombine the data

In the apply step, you can do things like apply a statistical function, filter out data, or transform the data.

Let's `groupby()` the wine_type in our wine `DataFrame`! Let's start with just `groupby()`, and then build it from there. This will produce a `DataFrame groupby` object.

In [64]:
wine.groupby('wine_type')

Not so interesting yet. This object has some attributes you can access. We can get lists of which rows are in which group by using the `.groups` attribute:

In [65]:
wine.groupby('wine_type').groups

{1: Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
             17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
             34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
             51, 52, 53, 54, 55, 56, 57, 58],
            dtype='int64'),
 2: Int64Index([ 59,  60,  61,  62,  63,  64,  65,  66,  67,  68,  69,  70,  71,
              72,  73,  74,  75,  76,  77,  78,  79,  80,  81,  82,  83,  84,
              85,  86,  87,  88,  89,  90,  91,  92,  93,  94,  95,  96,  97,
              98,  99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110,
             111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123,
             124, 125, 126, 127, 128, 129],
            dtype='int64'),
 3: Int64Index([130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142,
             143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155,
             156, 157, 158, 159, 160, 161, 162, 163, 164

The dataset was in order by `wine_type` to begin with, so that makes sense. To get just the keys, add the `.keys()` function to the end of that line.

In [66]:
wine.groupby('wine_type').groups.keys()

dict_keys([1, 2, 3])

Let's group our `auto_mpg` dataset by cylinders, just for contrast.

In [67]:
auto_mpg.groupby('cylinders').groups

{3: Int64Index([71, 111, 243, 334], dtype='int64'),
 4: Int64Index([ 14,  18,  19,  20,  21,  22,  23,  29,  30,  31,
             ...
             385, 388, 390, 391, 392, 393, 394, 395, 396, 397],
            dtype='int64', length=204),
 5: Int64Index([274, 297, 327], dtype='int64'),
 6: Int64Index([ 15,  16,  17,  24,  33,  34,  35,  36,  37,  45,  47,  48,  97,
              98,  99, 100, 101, 107, 113, 123, 125, 126, 127, 128, 133, 134,
             135, 152, 153, 154, 155, 160, 161, 162, 163, 164, 169, 174, 176,
             191, 192, 193, 194, 199, 200, 201, 202, 210, 211, 225, 226, 227,
             228, 241, 252, 253, 254, 256, 257, 258, 259, 260, 261, 263, 275,
             277, 280, 281, 283, 284, 306, 307, 316, 333, 341, 360, 361, 362,
             363, 365, 366, 386, 387, 389],
            dtype='int64'),
 8: Int64Index([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,
             ...
             286, 287, 288, 289, 290, 291, 292, 298, 300, 364],
            dtype='int6

You can see we have four observations with three cylinders, many more with four, and so on.

Going back to the wine example, let's apply an aggregate function. Let's generate the mean of all the other values and group them by `wine_class`.

In [68]:
wine.groupby('wine_type').mean()

abv  malic_acid       ash  alcalinity   magnesium  \
wine_type                                                            
1          13.744746    2.010678  2.455593   17.037288  106.338983   
2          12.278732    1.932676  2.244789   20.238028   94.549296   
3          13.153750    3.333750  2.437083   21.416667   99.312500   

           total_phenols  flavanoids  nonflavanoid_phenols  proanthocyacins  \
wine_type                                                                     
1               2.840169    2.982373              0.290000         1.899322   
2               2.258873    2.080845              0.363662         1.630282   
3               1.678750    0.781458              0.447500         1.153542   

              color       hue  dilution      proline  
wine_type                                             
1          5.528305  1.062034  3.157797  1115.711864  
2          3.086620  1.056282  2.785352   519.507042  
3          7.396250  0.682708  1.683542   629.895833

So, the mean `abv` for wine with type 1 is 13.74, type 2 is 12.27, type 3 is 13.15. The mean `malic_acid` for wine with type 1 is 2.01, and so on. So, with one line of code, we're able to apply a function to the entire dataset and see what's going on within different groups.

Selecting from a `groupby DataFrame` works the same way as selecting from any other `DataFrame`. Let's select the abv where `wine_type` is 2.

In [69]:
wine_type_mean = wine.groupby('wine_type').mean()

wine_type_mean.loc[2, 'abv']

12.278732394366198

It's also possible to apply multiple functions to the entire `DataFrame` using the `agg()` function. Let's get not only the mean, but the count and the standard deviation as well for each value in the `DataFrame`, still grouping by `wine_type`.

In [70]:
wine.groupby('wine_type').agg(['mean', 'count', 'std'])

abv                 malic_acid                       ash  \
                mean count       std       mean count       std      mean   
wine_type                                                                   
1          13.744746    59  0.462125   2.010678    59  0.688549  2.455593   
2          12.278732    71  0.537964   1.932676    71  1.015569  2.244789   
3          13.153750    48  0.530241   3.333750    48  1.087906  2.437083   

                          alcalinity  ...     color       hue                  \
          count       std       mean  ...       std      mean count       std   
wine_type                             ...                                       
1            59  0.227166  17.037288  ...  1.238573  1.062034    59  0.116483   
2            71  0.315467  20.238028  ...  0.924929  1.056282    71  0.202937   
3            48  0.184690  21.416667  ...  2.310942  0.682708    48  0.114441   

           dilution                      proline                    
               mean count       std         mean count         std  
wine_type                                                           
1          3.157797    59  0.357077  1115.711864    59  221.520767  
2          2.785352    71  0.496573   519.507042    71  157.211220  
3          1.683542    48  0.272111   629.895833    48  115.097043  

[3 rows x 39 columns]

It's also possible to run different functions on different columns. Let's get the mean for abv, the standard deviation for ash, and the sum of the values for hue. To do this, you'll need to create a dictionary with these functions, with the column names as the dictionary keys.

In [71]:
multiple_funcs = {'abv': 'std', 'ash': 'mean', 'hue': sum}

wine.groupby('wine_type').agg(multiple_funcs)

abv       ash     hue
wine_type                            
1          0.462125  2.455593  62.660
2          0.537964  2.244789  74.996
3          0.530241  2.437083  32.770

## Lesson: Groupby galore

Let's take this one step at a time.

In [73]:
# Let's group credit_approval by column G.
f.head()
f.groupby('G')

In [74]:
# Can you generate a list of all of the groups in the groupby object we just made?
f.groupby('G').groups

{'?': Int64Index([206, 270, 330, 456, 479, 539, 592, 601, 622], dtype='int64'),
 'bb': Int64Index([ 26,  27,  32,  48,  70,  71,  85,  95,  97, 100, 113, 126, 128,
             132, 133, 148, 159, 171, 180, 181, 186, 194, 195, 201, 209, 225,
             234, 236, 248, 318, 327, 334, 342, 343, 347, 378, 379, 380, 382,
             388, 391, 392, 402, 404, 413, 450, 458, 527, 532, 538, 541, 544,
             551, 595, 609, 613, 626, 640, 650],
            dtype='int64'),
 'dd': Int64Index([266, 364, 371, 375, 513, 521], dtype='int64'),
 'ff': Int64Index([ 39,  80, 111, 141, 212, 217, 261, 264, 278, 286, 294, 295, 296,
             300, 305, 306, 307, 321, 351, 359, 361, 362, 383, 397, 405, 410,
             415, 419, 428, 430, 435, 436, 437, 438, 441, 444, 445, 447, 448,
             463, 485, 488, 502, 533, 535, 536, 573, 585, 617, 627, 632, 633,
             649, 655, 669, 678, 687],
            dtype='int64'),
 'h': Int64Index([  1,   2,   6,   8,  10,  11,  18,  19,  23,  25,
      

In [76]:
# Let's use mean() on credit_approval_group to get the mean of our numeric values.
f.groupby('G').mean()

C         H         K             O
G                                              
?    1.106667  0.069444  0.000000     37.888889
bb   4.699068  3.034746  2.915254   1762.271186
dd   5.326667  0.569167  1.166667    485.000000
ff   5.377105  1.406316  1.859649    434.649123
h    5.368297  3.416667  3.036232    968.181159
j    3.218750  0.281250  1.500000    132.625000
n    4.593750  1.823750  2.750000    190.750000
o   11.667500  0.167500  0.000000  50060.000000
v    4.352845  1.813960  2.192982    792.992481
z   14.098750  8.218750  6.750000   2254.875000

In [77]:
# Let's see both the standard deviation and the sum of everything in credit_approval_group
f.groupby('G').agg(['mean','sum'])

C                   H                  K                  O  \
         mean       sum      mean      sum      mean  sum          mean   
G                                                                         
?    1.106667     9.960  0.069444    0.625  0.000000    0     37.888889   
bb   4.699068   277.245  3.034746  179.050  2.915254  172   1762.271186   
dd   5.326667    31.960  0.569167    3.415  1.166667    7    485.000000   
ff   5.377105   306.495  1.406316   80.160  1.859649  106    434.649123   
h    5.368297   740.825  3.416667  471.500  3.036232  419    968.181159   
j    3.218750    25.750  0.281250    2.250  1.500000   12    132.625000   
n    4.593750    18.375  1.823750    7.295  2.750000   11    190.750000   
o   11.667500    23.335  0.167500    0.335  0.000000    0  50060.000000   
v    4.352845  1736.785  1.813960  723.770  2.192982  875    792.992481   
z   14.098750   112.790  8.218750   65.750  6.750000   54   2254.875000   

            
       sum  
G           
?      341  
bb  103974  
dd    2910  
ff   24775  
h   133609  
j     1061  
n      763  
o   100120  
v   316404  
z    18039

In [80]:
# Let's see the count on column H, the sum on column C, and the mean on column O.
fun = {'H':'count', 'C':'sum','O':'mean'}

f.groupby('G').agg(fun)

H         C             O
G                              
?     9     9.960     37.888889
bb   59   277.245   1762.271186
dd    6    31.960    485.000000
ff   57   306.495    434.649123
h   138   740.825    968.181159
j     8    25.750    132.625000
n     4    18.375    190.750000
o     2    23.335  50060.000000
v   399  1736.785    792.992481
z     8   112.790   2254.875000

## Merge/join; or, how Pandas can be like SQL

In Pandas, it's possible to combine `DataFrames` and `Series` much like you would in SQL. For the examples in this section, we'll work with smaller `DataFrames` rather than our datasets. It's easier to provide proof of concept this way, as well as explain what's going on

Let's start by appending a row to a `DataFrame`. We can do that by passing in a dictionary to the append function, and setting `ignore_index` equal to `True`.

In [81]:
data = pd.DataFrame({'col1': ['i', 'love', 'pandas', 'so', 'much'],
        'col2': ['so', 'will', 'you', 'i', 'promise']})
data.append({'col1': 'dude', 'col2': 'dude'}, ignore_index=True)

col1     col2
0       i       so
1    love     will
2  pandas      you
3      so        i
4    much  promise
5    dude     dude

Appending a column is also easy. You can do that by setting a new column name equal to a list or a `Series`.

In [82]:
data['col3'] = ['how', 'do', 'you', 'like', 'oscon']
data

col1     col2   col3
0       i       so    how
1    love     will     do
2  pandas      you    you
3      so        i   like
4    much  promise  oscon

However, this will not work if your new column in a different length than the original `DataFrame`.

In [83]:
data['col4'] = ['I', 'am', 'too', 'short']
data

ValueError: Length of values does not match length of index

### Merge

You can `merge()` in different ways, just like joining in SQL. Let's look at an imaginary taco dataset:

In [84]:
tacos = pd.read_csv('../data/tacos.csv')
tacos

name      restaurant  number_of_tacos  score
0   Sarah      Taco Party                4    3.6
1  Georgi      Taco Mania                6    2.5
2   Sammy  Paradise Tacos               10    5.0
3   Peter      Taco Party                8    4.3
4     Rob      Taco Mania                8    3.4

Let's also look at an imaginary taco toppings dataset:

In [85]:
taco_toppings = pd.read_csv('../data/taco_toppings.csv')
taco_toppings

name favorite_topping least_favorite_topping corn_or_flour
0   Sammy            bacon                  slime          corn
1   Peter          avocado                   dirt         flour
2  Georgi         jalapeno                   dirt         flour
3   Sarah           cheese                 celery          corn
4     Rob            salsa                  slime         flour

Notice that we have a unique identifier in each dataset: the name column. We have the same five people. Let's merge these `DataFrames` together. You don't even need to pass the key to merge; `merge()` will automatically infer which key to use based on if it exists in both `DataFrames`. 

In [86]:
pd.merge(tacos, taco_toppings)

name      restaurant  number_of_tacos  score favorite_topping  \
0   Sarah      Taco Party                4    3.6           cheese   
1  Georgi      Taco Mania                6    2.5         jalapeno   
2   Sammy  Paradise Tacos               10    5.0            bacon   
3   Peter      Taco Party                8    4.3          avocado   
4     Rob      Taco Mania                8    3.4            salsa   

  least_favorite_topping corn_or_flour  
0                 celery          corn  
1                   dirt         flour  
2                  slime          corn  
3                   dirt         flour  
4                  slime         flour

By default, `merge()` performs a left outer join, which means it takes the key from the "left" `DataFrame` - the `DataFrame` that is passed in as the first parameter - and matches the right to it.

Generally speaking, full outer joins will join everything as a union, meaning that everything will be joined even if there are missing values; inner joins will join everything as an intersection, meaning that if a value does not appear in a row in a `DataFrame`, that row will be left out.

Let's look at a couple of other ways of merging. First, let's append a row to our tacos `DataFrame`.

In [87]:
tacos = tacos.append({'name': 'Dan', 'restaurant': 'Tres Carnes', 'number_of_tacos': 7, 'score': 3.8}, ignore_index=True)
tacos

name      restaurant  number_of_tacos  score
0   Sarah      Taco Party                4    3.6
1  Georgi      Taco Mania                6    2.5
2   Sammy  Paradise Tacos               10    5.0
3   Peter      Taco Party                8    4.3
4     Rob      Taco Mania                8    3.4
5     Dan     Tres Carnes                7    3.8

Now, let's do a full outer merge.

In [88]:
pd.merge(tacos, taco_toppings, how='outer')

name      restaurant  number_of_tacos  score favorite_topping  \
0   Sarah      Taco Party                4    3.6           cheese   
1  Georgi      Taco Mania                6    2.5         jalapeno   
2   Sammy  Paradise Tacos               10    5.0            bacon   
3   Peter      Taco Party                8    4.3          avocado   
4     Rob      Taco Mania                8    3.4            salsa   
5     Dan     Tres Carnes                7    3.8              NaN   

  least_favorite_topping corn_or_flour  
0                 celery          corn  
1                   dirt         flour  
2                  slime          corn  
3                   dirt         flour  
4                  slime         flour  
5                    NaN           NaN

You can see that the entire tacos `DataFrame` has been merged, even though 'Dan' does not exist in the `taco_toppings DataFrame`.

However, if we do the same thing and use a right outer join, we'll only use the keys from the `taco_toppings DataFrame` and Dan will be left out.

In [89]:
pd.merge(tacos, taco_toppings, how='right')

name      restaurant  number_of_tacos  score favorite_topping  \
0   Sarah      Taco Party                4    3.6           cheese   
1  Georgi      Taco Mania                6    2.5         jalapeno   
2   Sammy  Paradise Tacos               10    5.0            bacon   
3   Peter      Taco Party                8    4.3          avocado   
4     Rob      Taco Mania                8    3.4            salsa   

  least_favorite_topping corn_or_flour  
0                 celery          corn  
1                   dirt         flour  
2                  slime          corn  
3                   dirt         flour  
4                  slime         flour

### Join

The `join()` function gives you a way way to combine `DataFrames` without needing a key. `Taco_extra`, which contains data about chips and spiciness level, has no name column.

In [90]:
taco_extra = pd.read_csv('../data/taco_extra.csv')
taco_extra

chips spiciness
0   yes       hot
1    no      mild
2    no    medium
3   yes       hot
4   yes       hot

It's easy to join this to our taco `DataFrame`.

In [91]:
tacos.join(taco_extra)

name      restaurant  number_of_tacos  score chips spiciness
0   Sarah      Taco Party                4    3.6   yes       hot
1  Georgi      Taco Mania                6    2.5    no      mild
2   Sammy  Paradise Tacos               10    5.0    no    medium
3   Peter      Taco Party                8    4.3   yes       hot
4     Rob      Taco Mania                8    3.4   yes       hot
5     Dan     Tres Carnes                7    3.8   NaN       NaN

You can also specify how to join. The default is outer, but we can change it to inner and Dan will be left out again.

In [92]:
tacos.join(taco_extra, how='inner')

name      restaurant  number_of_tacos  score chips spiciness
0   Sarah      Taco Party                4    3.6   yes       hot
1  Georgi      Taco Mania                6    2.5    no      mild
2   Sammy  Paradise Tacos               10    5.0    no    medium
3   Peter      Taco Party                8    4.3   yes       hot
4     Rob      Taco Mania                8    3.4   yes       hot

It's possible to join more than two `DataFrames` at a time. Let's slice off the name column from taco_toppings.

In [93]:
taco_toppings_noname = taco_toppings.iloc[:, 1:]

taco_toppings_noname

favorite_topping least_favorite_topping corn_or_flour
0            bacon                  slime          corn
1          avocado                   dirt         flour
2         jalapeno                   dirt         flour
3           cheese                 celery          corn
4            salsa                  slime         flour

Joining this frame with tacos and taco_extra is as easy as chaining two joins together. Again, it's all an outer join, so even though there's no toppings or extra data for Dan, he's still included in the `DataFrame`.

In [94]:
tacos.join(taco_toppings_noname).join(taco_extra)

name      restaurant  number_of_tacos  score favorite_topping  \
0   Sarah      Taco Party                4    3.6            bacon   
1  Georgi      Taco Mania                6    2.5          avocado   
2   Sammy  Paradise Tacos               10    5.0         jalapeno   
3   Peter      Taco Party                8    4.3           cheese   
4     Rob      Taco Mania                8    3.4            salsa   
5     Dan     Tres Carnes                7    3.8              NaN   

  least_favorite_topping corn_or_flour chips spiciness  
0                  slime          corn   yes       hot  
1                   dirt         flour    no      mild  
2                   dirt         flour    no    medium  
3                 celery          corn   yes       hot  
4                  slime         flour   yes       hot  
5                    NaN           NaN   NaN       NaN

## Lesson: Let's merge some dataframes!

In [107]:
# Can you merge following DataFrames together?
pizza = pd.read_csv('../data/pizza.csv')
pizza_toppings = pd.read_csv('../data/pizza_toppings.csv')

# Merge them here
pd.merge(pizza,pizza_toppings)

name pizza_style  number_of_slices flavored_crust favorite_topping  \
0     Will    new_york                 5            yes        pepperoni   
1     Lane    new_york                 4             no        anchovies   
2     Dave     chicago                 4             no     green_pepper   
3  Shannon     chicago                 6             no            onion   
4    Ethan     chicago                 3            yes              ham   

  least_favorite  
0      anchovies  
1      pineapple  
2      anchovies  
3      anchovies  
4      bbq_sauce

In [108]:
pizza

name pizza_style  number_of_slices flavored_crust
0     Will    new_york                 5            yes
1     Lane    new_york                 4             no
2     Dave     chicago                 4             no
3  Shannon     chicago                 6             no
4    Ethan     chicago                 3            yes

In [109]:
pizza_toppings

name favorite_topping least_favorite
0     Will        pepperoni      anchovies
1     Lane        anchovies      pineapple
2     Dave     green_pepper      anchovies
3  Shannon            onion      anchovies
4    Ethan              ham      bbq_sauce
5    Peggy        pineapple      anchovies

In [110]:
# Let's inner merge those DataFrames

pd.merge(pizza,pizza_toppings,how='inner')

name pizza_style  number_of_slices flavored_crust favorite_topping  \
0     Will    new_york                 5            yes        pepperoni   
1     Lane    new_york                 4             no        anchovies   
2     Dave     chicago                 4             no     green_pepper   
3  Shannon     chicago                 6             no            onion   
4    Ethan     chicago                 3            yes              ham   

  least_favorite  
0      anchovies  
1      pineapple  
2      anchovies  
3      anchovies  
4      bbq_sauce

In [111]:
# Let's join pizza to another dataset, pizza_extra
pizza_extra = pd.read_csv('../data/pizza_extra.csv')
pizza_extra 

meal heating_method
0     dinner           oven
1      lunch      microwave
2  breakfast           oven
3  breakfast      microwave

In [112]:
# Let's only join them together where all the data is present
pizza.join(pizza_extra)

name pizza_style  number_of_slices flavored_crust       meal  \
0     Will    new_york                 5            yes     dinner   
1     Lane    new_york                 4             no      lunch   
2     Dave     chicago                 4             no  breakfast   
3  Shannon     chicago                 6             no  breakfast   
4    Ethan     chicago                 3            yes        NaN   

  heating_method  
0           oven  
1      microwave  
2           oven  
3      microwave  
4            NaN

In [115]:
# Can you join all three dataframes together, first by merging pizza and pizza_toppings, then joining that to pizza_extra?
pizza_top_noname = pizza_toppings.iloc[:,1:]
pizza.join(pizza_top_noname).join(pizza_extra)

name pizza_style  number_of_slices flavored_crust favorite_topping  \
0     Will    new_york                 5            yes        pepperoni   
1     Lane    new_york                 4             no        anchovies   
2     Dave     chicago                 4             no     green_pepper   
3  Shannon     chicago                 6             no            onion   
4    Ethan     chicago                 3            yes              ham   

  least_favorite       meal heating_method  
0      anchovies     dinner           oven  
1      pineapple      lunch      microwave  
2      anchovies  breakfast           oven  
3      anchovies  breakfast      microwave  
4      bbq_sauce        NaN            NaN

## Pivoting

You can pivot in Pandas just like you would in Excel. `pivot_table()` takes in four requires parameters: the `DataFrame`, the column to use for the index, the column to use for the columns, and the column to use for the values. `pivot_table()` also has an `aggfunc` parameter that defaults to the mean of the values, but you can pass in other functions, just as we did in the `agg()` function before.

Let's look at the mean weight per model number and number of cylinders combination.

In [116]:
pd.pivot_table(auto_mpg, values='weight', index='model', columns='cylinders')

cylinders       3            4       5            6            8
model                                                           
70            NaN  2292.571429     NaN  2710.500000  3940.055556
71            NaN  2056.384615     NaN  3171.875000  4537.714286
72         2330.0  2382.642857     NaN          NaN  4228.384615
73         2124.0  2338.090909     NaN  2917.125000  4279.050000
74            NaN  2151.466667     NaN  3320.000000  4438.400000
75            NaN  2489.250000     NaN  3398.333333  4108.833333
76            NaN  2306.600000     NaN  3349.600000  4064.666667
77         2720.0  2205.071429     NaN  3383.000000  4177.500000
78            NaN  2296.764706  2830.0  3314.166667  3563.333333
79            NaN  2357.583333  3530.0  3025.833333  3862.900000
80         2420.0  2360.080000  2950.0  3145.500000          NaN
81            NaN  2275.476190     NaN  3093.571429  3725.000000
82            NaN  2402.321429     NaN  2931.666667          NaN

If a cell contains NaN, it means that that combination doesn't exist within the `DataFrame`.

We can pass in multiple column names to the rows and cols parameters. This creates a `multiindex`.

If we add the origin column to our pivot table, we can look at the average weight of all of the model/origin combinations against the number of cylinders the cars have.

In [117]:
pd.pivot_table(auto_mpg, values='weight', index=['model', 'origin'], columns='cylinders')

cylinders          3            4       5            6            8
model origin                                                       
70    1          NaN          NaN     NaN  2710.500000  3940.055556
      2          NaN  2309.200000     NaN          NaN          NaN
      3          NaN  2251.000000     NaN          NaN          NaN
71    1          NaN  2178.600000     NaN  3171.875000  4537.714286
      2          NaN  2024.000000     NaN          NaN          NaN
      3          NaN  1936.000000     NaN          NaN          NaN
72    1          NaN  2263.800000     NaN          NaN  4228.384615
      2          NaN  2573.200000     NaN          NaN          NaN
      3       2330.0  2293.000000     NaN          NaN          NaN
73    1          NaN  2355.500000     NaN  2932.857143  4279.050000
      2          NaN  2335.714286     NaN          NaN          NaN
      3       2124.0  2329.000000     NaN  2807.000000          NaN
74    1          NaN  2372.666667     NaN  3320.000000  4438.400000
      2          NaN  2139.333333     NaN          NaN          NaN
      3          NaN  2053.000000     NaN          NaN          NaN
75    1          NaN  2615.500000     NaN  3398.333333  4108.833333
      2          NaN  2571.166667     NaN          NaN          NaN
      3          NaN  2303.250000     NaN          NaN          NaN
76    1          NaN  2318.200000     NaN  3343.250000  4064.666667
      2          NaN  2438.285714     NaN  3820.000000          NaN
      3          NaN  1980.000000     NaN  2930.000000          NaN
77    1          NaN  2346.000000     NaN  3525.000000  4177.500000
      2          NaN  2138.750000     NaN          NaN          NaN
      3       2720.0  2060.000000     NaN  2815.000000          NaN
78    1          NaN  2417.500000     NaN  3322.000000  3563.333333
      2          NaN  2256.666667  2830.0  3275.000000          NaN
      3          NaN  2221.250000     NaN          NaN          NaN
79    1          NaN  2435.857143     NaN  3025.833333  3862.900000
      2          NaN  2415.000000  3530.0          NaN          NaN
      3          NaN  1997.500000     NaN          NaN          NaN
80    1          NaN  2729.333333     NaN  3381.000000          NaN
      2          NaN  2272.750000  2950.0          NaN          NaN
      3       2420.0  2222.181818     NaN  2910.000000          NaN
81    1          NaN  2330.625000     NaN  3166.250000  3725.000000
      2          NaN  2580.000000     NaN  3160.000000          NaN
      3          NaN  2140.000000     NaN  2915.000000          NaN
82    1          NaN  2585.882353     NaN  2931.666667          NaN
      2          NaN  2055.000000     NaN          NaN          NaN
      3          NaN  2132.777778     NaN          NaN          NaN

You can apply different aggregate functions to a pivot table. Let's look at the total weight per model/cylinder combination.

In [118]:
pd.pivot_table(auto_mpg, values='weight', index='model', columns='cylinders', aggfunc='sum')

cylinders       3        4       5        6        8
model                                               
70            NaN  16048.0     NaN  10842.0  70921.0
71            NaN  26733.0     NaN  25375.0  31764.0
72         2330.0  33357.0     NaN      NaN  54969.0
73         2124.0  25719.0     NaN  23337.0  85581.0
74            NaN  32272.0     NaN  23240.0  22192.0
75            NaN  29871.0     NaN  40780.0  24653.0
76            NaN  34599.0     NaN  33496.0  36582.0
77         2720.0  30871.0     NaN  16915.0  33420.0
78            NaN  39045.0  2830.0  39770.0  21380.0
79            NaN  28291.0  3530.0  18155.0  38629.0
80         2420.0  59002.0  2950.0   6291.0      NaN
81            NaN  47785.0     NaN  21655.0   3725.0
82            NaN  67265.0     NaN   8795.0      NaN

## Lesson: let's pivot!

In [120]:
# Create a pivot_table for credit_approval with column A as the index, column J as the columns, and column H as the values.
pd.pivot_table(f, values='H', index='A', columns='J')

J         f         t
A                    
?  0.750000  2.416667
a  1.405370  2.188725
b  1.673291  3.578658

In [121]:
# Now, change the aggfunc to the standard deviation.
pd.pivot_table(f, values='H', index='A', columns='J',aggfunc='std')

J         f         t
A                    
?  1.203482  2.036132
a  2.316387  2.736864
b  2.660253  4.531032

In [128]:
# Finally, can you come up with your own pivot_table?
f.info()

pd.pivot_table(f, values='C', index='A', columns='I')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 690 entries, 0 to 689
Data columns (total 16 columns):
A    690 non-null object
B    690 non-null object
C    690 non-null float64
D    690 non-null object
E    690 non-null object
F    690 non-null object
G    690 non-null object
H    690 non-null float64
I    690 non-null object
J    690 non-null object
K    690 non-null int64
L    690 non-null object
M    690 non-null object
N    690 non-null object
O    690 non-null int64
P    690 non-null object
dtypes: float64(2), int64(2), object(12)
memory usage: 86.4+ KB


I         f          t
A                     
?  1.608500  10.125000
a  3.757604   6.180132
b  3.452713   5.763122

# For those using IPython Notebook/Wakari/NBViewer: Go to the [data_analysis](data_analysis.ipynb) notebook!

# For those using code files, go to data_analysis.py!